README.rst
``pprint++``: a drop-in replacement for ``pprint`` that's actually pretty
=========================================================================

This is a continuation of `pprintpp <https://github.com/wolever/pprintpp>`_,
by `David Wolever <https://github.com/wolever>`_, which seems to have been
discontinued by the original writer. Currently, under development still.

Installation
------------


``pprint++`` can be installed using ``pip`` or ``easy_install``::

    $ pip install pprintpp
    - OR -
    $ easy_install pprintpp

Usage
-----

``pprint++`` can be used in three ways:

1. Through the separate ``pp`` package::

    $ pip install pp-ez
    $ python
    ...
    >>> import pp
    >>> pp(["Hello", "world"])
    ["Hello", "world"]

   For more, see https://pypi.python.org/pypi/pp-ez

2. As a command-line program, which will read Python literals from standard in
   and pretty-print them::

    $ echo "{'hello': 'world'}" | pypprint
    {'hello': 'world'}

3. As an `ipython <https://github.com/ipython/ipython>`_ extension::

    In [1]: %load_ext pprintpp
   
   This will use pprintpp for ipython's output.
   
   To load this extension when ipython starts, put the previous line in your `startup file <https://ipython.org/ipython-doc/1/config/overview.html#startup-files>`_.
   
   You can change the indentation level like so::
    
    In [2]: %config PPrintPP.indentation = 4 

4. To monkeypatch ``pprint``::

    >>> import pprintpp
    >>> pprintpp.monkeypatch()
    >>> import pprint
    >>> pprint.pprint(...)

   Note: the original ``pprint`` module will be available with ``import
   pprint_original``. Additionally, a warning will be issued if ``pprint`` has
   already been imported. This can be suppressed by passing ``quiet=True``.

5. And, if you *really* want, it can even be imported as a regular module:

   >>> import pprintpp
   >>> pprintpp.pprint(...)


Usability Protips
-----------------

``pp``
~~~~~~

For bonus code aesthetics, ``pprintpp.pprint`` can be imported as ``pp``:

.. code:: pycon

    >>> from pprintpp import pprint as pp
    >>> pp(...)

And if that is just too many letters, the ``pp-ez`` package can be installed
from PyPI, ensuring that pretty-printing is never more than an ``import pp``
away::

    $ pip install pp-ez
    $ python
    ...
    >>> import pp
    >>> pp(["Hello", "world"])
    ["Hello", "world"]

For more, see https://pypi.python.org/pypi/pp-ez


Why is it prettier?
-------------------

Unlike ``pprint``, ``pprint++`` strives to emit a readable, largely
PEP8-compliant, representation of its input.

It also has explicit support for: the ``collections`` module (``defaultdict``
and ``Counter``) and ``numpy`` arrays:

.. code:: pycon

    >>> import numpy as np
    >>> from collections import defaultdict, Counter
    >>> pprint([np.array([[1,2],[3,4]]), defaultdict(int, {"foo": 1}), Counter("aaabbc")])
    [
        array([[1, 2],
               [3, 4]]),
        defaultdict(<type 'int'>, {'foo': 1}),
        Counter({'a': 3, 'b': 2, 'c': 1}),
    ]

Unicode characters, when possible, will be printed un-escaped. This is done by
checking both the output stream's encoding (defaulting to ``utf-8``) and the
character's Unicode category. An effort is made to print only characters which
will be visually unambiguous: letters and numbers will be printed un-escaped,
spaces, combining characters, and control characters will be escaped:

.. code:: pycon

    >>> unistr = u"\xe9e\u0301"
    >>> print unistr
    éé
    >>> pprint(unistr)
    u'ée\u0301'

The output stream's encoding will be considered too:

.. code:: pycon

    >>> import io
    >>> stream = io.BytesIO()
    >>> stream.encoding = "ascii"
    >>> pprint(unistr, stream=stream)
    >>> print stream.getvalue()
    u'\xe9e\u0301'

Subclassess of built-in collection types which don't define a new ``__repr__``
will have their class name explicitly added to their repr. For example:

.. code:: pycon

    >>> class MyList(list):
    ...     pass
    ...
    >>> pprint(MyList())
    MyList()
    >>> pprint(MyList([1, 2, 3]))
    MyList([1, 2, 3])

Note that, as you might expect, custom ``__repr__`` methods will be respected:

.. code:: pycon

    >>> class MyList(list):
    ...     def __repr__(self):
    ...         return "custom repr!"
    ...
    >>> pprint(MyList())
    custom repr!

**Note**: ``pprint++`` is still under development, so the format *will* change
and improve over time.

Example
~~~~~~~

With ``printpp``:

.. code:: pycon

    >>> import pprintpp
    >>> pprintpp.pprint(["Hello", np.array([[1,2],[3,4]])])
    [
        'Hello',
        array([[1, 2],
               [3, 4]]),
    ]
    >>> pprintpp.pprint(tweet)
    {
        'coordinates': None,
        'created_at': 'Mon Jun 27 19:32:19 +0000 2011',
        'entities': {
            'hashtags': [],
            'urls': [
                {
                    'display_url': 'tumblr.com/xnr37hf0yz',
                    'expanded_url': 'http://tumblr.com/xnr37hf0yz',
                    'indices': [107, 126],
                    'url': 'http://t.co/cCIWIwg',
                },
            ],
            'user_mentions': [],
        },
        'place': None,
        'source': '<a href="http://www.tumblr.com/" rel="nofollow">Tumblr</a>',
        'truncated': False,
        'user': {
            'contributors_enabled': True,
            'default_profile': False,
            'entities': {'hashtags': [], 'urls': [], 'user_mentions': []},
            'favourites_count': 20,
            'id_str': '6253282',
            'profile_link_color': '0094C2',
        },
    }

Without ``printpp``::

    >>> import pprint
    >>> import numpy as np
    >>> pprint.pprint(["Hello", np.array([[1,2],[3,4]])])
    ['Hello', array([[1, 2],
           [3, 4]])]
    >>> tweet = {'coordinates': None, 'created_at': 'Mon Jun 27 19:32:19 +0000 2011', 'entities': {'hashtags': [], 'urls': [{'display_url': 'tumblr.com/xnr37hf0yz', 'expanded_url': 'http://tumblr.com/xnr37hf0yz', 'indices': [107, 126], 'url': 'http://t.co/cCIWIwg'}], 'user_mentions': []}, 'place': None, 'source': '<a href="http://www.tumblr.com/" rel="nofollow">Tumblr</a>', 'truncated': False, 'user': {'contributors_enabled': True, 'default_profile': False, 'entities': {'hashtags': [], 'urls': [], 'user_mentions': []}, 'favourites_count': 20, 'id_str': '6253282', 'profile_link_color': '0094C2'}} 
    >>> pprint.pprint(tweet)
    {'coordinates': None,
     'created_at': 'Mon Jun 27 19:32:19 +0000 2011',
     'entities': {'hashtags': [],
                  'urls': [{'display_url': 'tumblr.com/xnr37hf0yz',
                            'expanded_url': 'http://tumblr.com/xnr37hf0yz',
                            'indices': [107, 126],
                            'url': 'http://t.co/cCIWIwg'}],
                  'user_mentions': []},
     'place': None,
     'source': '<a href="http://www.tumblr.com/" rel="nofollow">Tumblr</a>',
     'truncated': False,
     'user': {'contributors_enabled': True,
              'default_profile': False,
              'entities': {'hashtags': [], 'urls': [], 'user_mentions': []},
              'favourites_count': 20,
              'id_str': '6253282',
              'profile_link_color': '0094C2'}}


CHANGELOG.txt
0.5.0 (unreleased)
    * Update documentation to point to `pprintpp2` instead.
    * Remove support for older Python versions (3.7+ from now on).
    * Add `sort_dict` option (thanks @gergelyk).
    * Use `pytest` instead of `nose` (thanks @pgajdos).
    * Use GitHub Actions instead of Travis.
    * Linting: `isort`, `black`, `flake8`.
      Wrapping at 100 lines from the previous 80.
    * Security: `safety` and `bandit`.
    * Other file and folder restructuring.

0.4.0
    * Add IPython plugin (thanks @roee30;
      https://github.com/wolever/pprintpp/pull/20)

0.3.0
    * Add support for subclasses of built-in collections and collections
      module (ex, if no __repr__ is defined, a subclass of dict will be
      repr'd ``MyDictSubclass({...})``).
    * Add support for ``collections.OrderedDict``
    * Add ``quiet=True`` flag to monkeypatch
    * Fixed crash on non-hashable instancemethods
      (https://github.com/wolever/pprintpp/issues/18)

0.2.3 (2015-06-01)
    * Oops, fixed setup.py so the module is actually push'd to PyPI (thanks
      @OldhamMade).
    * pprintpp is now a package, not a single .py file.

0.2.2 (2015-05-14)
    * Add Python 2.6 compatibility (pprintpp will work without the
      `collections` module; thanks @ulope)
    * Fix set/frozenset bug on PyPy
    * Fix sorting of heterogeneous containers on Py3

0.2.1 (2014-04-22)
    * Some minor bugfixes

0.2.0 (2014-04-22)
    * Add Python 3 support

0.1.3 (2014-04-03)
    * Fix bug in printing Unicode strings when locale is imported
    * Improve Unicode printing so a wider range of Unicode characters will be
      printed directly.

0.1.2 (2014-03-30)
    * Add support for collections.Counter and collections.defaultdict

0.1.1 (2014-03-30)
    * Proper numpy support, fix sorting bug

0.1.0 (2014-03-30)
    * Initial release.


test.py
from __future__ import print_function

import sys
from contextlib import redirect_stdout
import io
import pytest

sys.path.append("pp/")
import pp
import pprintpp as p
from pprintpp import Counter, defaultdict, OrderedDict

def test_pp():
    expected = "['hello', 'world']"
    f = io.StringIO()
    with redirect_stdout(f):
        pp(["hello", "world"])
    actual = f.getvalue().rstrip("\n")
    assert actual == expected

def test_pp_print():
    expected = "'stuff'"
    f = io.StringIO()
    with redirect_stdout(f):
        pp.pprint("stuff")
    actual = f.getvalue().rstrip("\n")
    assert actual == expected

def test_fmt():
    expected = "'asdf'\n'stuff'"
    f = io.StringIO()
    with redirect_stdout(f):
        print(pp.pformat("asdf"))
        print(pp.fmt("stuff"))
    actual = f.getvalue().rstrip("\n")
    assert actual == expected

def test_module_like():
     print(dir(pp))
     print(repr(pp))

uni_safe = "\xe9 \u6f02 \u0e4f \u2661"
uni_unsafe = "\u200a \u0302 \n"
slashed = lambda s: u"'%s'" %(
    s.encode("ascii", "backslashreplace").decode("ascii").replace("\n", "\\n")
)

@pytest.mark.skip('fix')
@pytest.mark.parametrize("input,expected,encoding", [
    (uni_safe, "%s'%s'" %('', uni_safe), "utf-8"),
    (uni_unsafe, slashed(uni_unsafe), "utf-8"),
    (uni_unsafe, slashed(uni_unsafe), "ascii"),
    ("\U0002F9B2", slashed("\U0002F9B2"), "ascii")
])
def test_unicode(input, expected, encoding):
    stream = p.TextIO(encoding=encoding)
    p.pprint(input, stream=stream)
    assert stream.getvalue().rstrip("\n") == expected

test_back_and_forth_data = [
    "'\\'\"'",
    '"\'"',
    "'\"'",
    "frozenset(['a', 'b', 'c'])",
    "set([None, 1, 'a'])",
    "[]",
    "[1]",
    "{}",
    "{1: 1}",
    "set()",
    "set([1])",
    "frozenset()",
    "frozenset([1])",
    "()",
    "(1, )",
    "MyDict({})",
    "MyDict({1: 1})",
    "MyList([])",
    "MyList([1])",
    "MyTuple(())",
    "MyTuple((1, ))",
    "MySet()",
    "MySet([1])",
    "MyFrozenSet()",
    "MyFrozenSet([1])",
    "Counter()",
    "Counter({1: 1})",
    "OrderedDict()",
    "OrderedDict([(1, 1), (5, 5), (2, 2)])",
    "MyOrderedDict()",
    "MyOrderedDict([(1, 1)])",
    "MyCounter()",
    "MyCounter({1: 1})",
    "MyCounterWithRepr('dummy')",
]

class MyDict(dict):
    pass

class MyList(list):
    pass

class MyTuple(tuple):
    pass

class MySet(set):
    pass

class MyFrozenSet(frozenset):
    pass

class MyOrderedDict(p.OrderedDict):
    pass

class MyDefaultDict(p.defaultdict):
    pass

class MyCounter(p.Counter):
    pass

class MyCounterWithRepr(p.Counter):
    def __repr__(self):
        return "MyCounterWithRepr('dummy')"

@pytest.mark.skip('fix')
@pytest.mark.parametrize("expected", test_back_and_forth_data)
def test_back_and_forth(expected):
    input = eval(expected)
    stream = p.TextIO()
    p.pprint(input, stream=stream)
    assert stream.getvalue().rstrip("\n") == expected

test_expected_input_data = [
    ("defaultdict(%r, {})" %(int, ), defaultdict(int)),
    ("defaultdict(%r, {1: 1})" %(int, ), defaultdict(int, [(1, 1)])),
    ("MyDefaultDict(%r, {})" %(int, ), MyDefaultDict(int)),
    ("MyDefaultDict(%r, {1: 1})" %(int, ), MyDefaultDict(int, [(1, 1)])),
]

@pytest.mark.skip('fix')
@pytest.mark.parametrize("expected,input", test_expected_input_data)
def test_expected_input(expected, input):
    stream = p.TextIO()
    p.pprint(input, stream=stream)
    assert stream.getvalue().rstrip("\n") == expected


def test_unhashable_repr():
    # In Python 3, C extensions can define a __repr__ method which is an
    # instance of `instancemethod`, which is unhashable. It turns out to be
    # spectacularly difficult to create an `instancemethod` and attach it to
    # a type without using C... so we'll simulate it using a more explicitly
    # unhashable type.
    # See also: http://stackoverflow.com/q/40876368/71522

    class UnhashableCallable(object):
        __hash__ = None

        def __call__(self):
            return "some-repr"

    class MyCls(object):
        __repr__ = UnhashableCallable()

    obj = MyCls()
    assert p.pformat(obj) == "some-repr"





pprintpp/__init__.py
from __future__ import print_function

import ast
import io
import os
import sys
import unicodedata
import warnings
from collections import Counter, OrderedDict, defaultdict

from .safesort import safesort

__all__ = ["pprint", "pformat", "isreadable", "isrecursive", "saferepr", "PrettyPrinter"]


def chr_to_ascii(c):
    return ascii(c)[1:-1]


def _sorted(iterable):
    try:
        return sorted(iterable)
    except TypeError:
        return safesort(iterable)


class TextIO(io.TextIOWrapper):
    def __init__(self, encoding=None):
        io.TextIOWrapper.__init__(self, io.BytesIO(), encoding=encoding)

    def getvalue(self):
        self.flush()
        return self.buffer.getvalue().decode(self.encoding)


# pprintpp will make an attempt to print as many Unicode characters as is
# safely possible. It will use the character category along with this table to
# determine whether or not it is safe to print a character. In this context,
# "safety" is defined as "the character will appear visually distinct" -
# combining characters, spaces, and other things which could be visually
# ambiguous are repr'd, others will be printed. I made this table mostly by
# hand, mostly guessing, so please file bugs.
# Source: http://www.unicode.org/reports/tr44/#GC_Values_Table
# fmt: off
unicode_printable_categories = {
    "Lu": 1,  # Uppercase_Letter	an uppercase letter
    "Ll": 1,  # Lowercase_Letter	a lowercase letter
    "Lt": 1,  # Titlecase_Letter	a digraphic character, with first part uppercase
    "LC": 1,  # Cased_Letter	Lu | Ll | Lt
    "Lm": 0,  # Modifier_Letter	a modifier letter
    "Lo": 1,  # Other_Letter	other letters, including syllables and ideographs
    "L":  1,  # Letter	Lu | Ll | Lt | Lm | Lo
    "Mn": 0,  # Nonspacing_Mark	a nonspacing combining mark (zero advance width)
    "Mc": 0,  # Spacing_Mark	a spacing combining mark (positive advance width)
    "Me": 0,  # Enclosing_Mark	an enclosing combining mark
    "M":  1,  # Mark	Mn | Mc | Me
    "Nd": 1,  # Decimal_Number	a decimal digit
    "Nl": 1,  # Letter_Number	a letterlike numeric character
    "No": 1,  # Other_Number	a numeric character of other type
    "N":  1,  # Number	Nd | Nl | No
    "Pc": 1,  # Connector_Punctuation	a connecting punctuation mark, like a tie
    "Pd": 1,  # Dash_Punctuation	a dash or hyphen punctuation mark
    "Ps": 1,  # Open_Punctuation	an opening punctuation mark (of a pair)
    "Pe": 1,  # Close_Punctuation	a closing punctuation mark (of a pair)
    "Pi": 1,  # Initial_Punctuation	an initial quotation mark
    "Pf": 1,  # Final_Punctuation	a final quotation mark
    "Po": 1,  # Other_Punctuation	a punctuation mark of other type
    "P":  1,  # Punctuation	Pc | Pd | Ps | Pe | Pi | Pf | Po
    "Sm": 1,  # Math_Symbol	a symbol of mathematical use
    "Sc": 1,  # Currency_Symbol	a currency sign
    "Sk": 1,  # Modifier_Symbol	a non-letterlike modifier symbol
    "So": 1,  # Other_Symbol	a symbol of other type
    "S":  1,  # Symbol	Sm | Sc | Sk | So
    "Zs": 0,  # Space_Separator	a space character (of various non-zero widths)
    "Zl": 0,  # Line_Separator	U+2028 LINE SEPARATOR only
    "Zp": 0,  # Paragraph_Separator	U+2029 PARAGRAPH SEPARATOR only
    "Z":  1,  # Separator	Zs | Zl | Zp
    "Cc": 0,  # Control	a C0 or C1 control code
    "Cf": 0,  # Format	a format control character
    "Cs": 0,  # Surrogate	a surrogate code point
    "Co": 0,  # Private_Use	a private-use character
    "Cn": 0,  # Unassigned	a reserved unassigned code point or a noncharacter
    "C":  0,  # Other	Cc | Cf | Cs | Co | Cn
}
# fmt: on

ascii_table = dict((chr(i), chr_to_ascii(chr(i))) for i in range(255))


def pprint(object, stream=None, indent=4, width=80, depth=None, sort_dicts=True):
    """Pretty-print a Python object to a stream [default is sys.stdout]."""
    printer = PrettyPrinter(
        stream=stream, indent=indent, width=width, depth=depth, sort_dicts=sort_dicts
    )
    printer.pprint(object)


def pformat(object, indent=4, width=80, depth=None, sort_dicts=True):
    """Format a Python object into a pretty-printed representation."""
    return PrettyPrinter(indent=indent, width=width, depth=depth, sort_dicts=sort_dicts).pformat(
        object
    )


def saferepr(object):
    """Version of repr() which can handle recursive data structures."""
    return PrettyPrinter().pformat(object)


def isreadable(object):
    """Determine if saferepr(object) is readable by eval()."""
    return PrettyPrinter().isreadable(object)


def isrecursive(object):
    """Determine if object requires a recursive representation."""
    return PrettyPrinter().isrecursive(object)


def console(argv=None):
    if argv is None:
        argv = sys.argv
    if len(argv) != 1:
        name = argv[0]
        if name.startswith("/"):
            name = os.path.basename(name)
        print("Usage: %s" % (argv[0],))
        print()
        print("Pipe Python literals into %s to pretty-print them" % (argv[0],))
        return 1
    obj = ast.literal_eval(sys.stdin.read().strip())
    pprint(obj)
    return 0


def monkeypatch(mod=None, quiet=False):
    if "pprint" in sys.modules and not quiet:
        warnings.warn("'pprint' has already been imported; monkeypatching won't work everywhere.")
    import pprint

    sys.modules["pprint_original"] = pprint
    sys.modules["pprint"] = mod or sys.modules["pprintpp"]


class PPrintSharedState(object):
    recursive = False
    readable = True
    cur_line_length = 0

    def clone(self):
        new = type(self)()
        new.__dict__.update(self.__dict__)
        return new


class PPrintState(object):
    indent = 4
    level = 0
    max_width = 80
    max_depth = None
    sort_dicts = True
    stream = None
    context = None
    write_constrain = None

    class WriteConstrained(Exception):
        pass

    def __init__(self, **attrs):
        self.__dict__.update(attrs)
        self.s = PPrintSharedState()

    def assert_sanity(self):
        assert self.indent >= 0, "indent must be >= 0"
        assert self.max_depth is None or self.max_depth > 0, "depth must be > 0"
        assert self.max_width, "width must be != 0"

    def replace(self, **attrs):
        new_state = type(self)()
        new_state.__dict__.update(self.__dict__)
        new_state.__dict__.update(attrs)
        new_state.context = dict(new_state.context)
        new_state.s = self.s
        return new_state

    def clone(self, clone_shared=False):
        new = self.replace()
        if clone_shared:
            new.s = self.s.clone()
        return new

    def write(self, data):
        if self.write_constrain is not None:
            self.write_constrain -= len(data)
            if self.write_constrain < 0:
                raise self.WriteConstrained

        if isinstance(data, bytes):
            data = data.decode("latin1")
        self.stream.write(data)
        nl_idx = data.rfind("\n")
        if nl_idx < 0:
            self.s.cur_line_length += len(data)
        else:
            self.s.cur_line_length = len(data) - (nl_idx + 1)

    def get_indent_string(self):
        return (self.level * self.indent) * " "


def _mk_open_close_empty_dict(type_tuples):
    """
    Generates a dictionary mapping either ``cls.__repr__`` xor ``cls`` to
    a tuple of ``(container_type, repr_open, repr_close, repr_empty)`` (see
    ``PrettyPrinter._open_close_empty`` for examples).

    Using either ``cls.__repr__`` xor ``cls`` is important because some
    types (specifically, ``set`` and ``frozenset`` on PyPy) share a
    ``__repr__``. When we are determining how to repr an object, the type
    is first checked, then if it's not found ``type.__repr__`` is checked.

    Note that ``__repr__`` is used so that trivial subclasses will behave
    sensibly."""

    res = {}
    for (cls, open_close_empty) in type_tuples:
        if cls.__repr__ in res:
            res[cls] = (cls,) + open_close_empty
        else:
            res[cls.__repr__] = (cls,) + open_close_empty
    return res


class PrettyPrinter(object):
    def __init__(self, indent=4, width=80, depth=None, stream=None, sort_dicts=True):
        """
        Handle pretty printing operations onto a stream using a set of
        configured parameters.

        indent
            Number of spaces to indent for each level of nesting.

        width
            Attempted maximum number of columns in the output.

        depth
            The maximum depth to print out nested structures.

        stream
            The desired output stream.  If omitted (or false), the standard
            output stream available at construction will be used.

        sort_dicts
            If `True`, dictionaries will be formatted with their keys sorted,
            otherwise they will display in the order as returned by `items`
            method.

        """
        self.get_default_state = lambda: PPrintState(
            indent=int(indent),
            max_width=int(width),
            sort_dicts=sort_dicts,
            stream=stream or sys.stdout,
            context={},
        )
        self.get_default_state().assert_sanity()

    def pprint(self, object, state=None):
        state = state or self.get_default_state()
        self._format(object, state)
        state.write("\n")

    def pformat(self, object, state=None):
        sio = TextIO()
        state = state or self.get_default_state()
        state = state.replace(stream=sio)
        self._format(object, state)
        return sio.getvalue()

    def isrecursive(self, object):
        state = self.get_default_state()
        self._format(object, state)
        return state.s.recursive

    def isreadable(self, object):
        state = self.get_default_state()
        self._format(object, state)
        return state.s.readable and not state.s.recursive

    _open_close_empty = _mk_open_close_empty_dict(
        [
            (dict, ("dict", "{", "}", "{}")),
            (list, ("list", "[", "]", "[]")),
            (tuple, ("tuple", "(", ")", "()")),
            (set, ("set", "__PP_TYPE__([", "])", "__PP_TYPE__()")),
            (frozenset, ("set", "__PP_TYPE__([", "])", "__PP_TYPE__()")),
            (Counter, ("dict", "__PP_TYPE__({", "})", "__PP_TYPE__()")),
            (defaultdict, ("dict", None, "})", None)),
            (OrderedDict, ("odict", "__PP_TYPE__([", "])", "__PP_TYPE__()")),
        ]
    )

    def _format_nested_objects(self, object, state, typeish=None):
        objid = id(object)
        state.level += 1
        state.context[objid] = 1
        try:
            # First, try to fit everything on one line. For simplicity, assume
            # that it takes three characters to close the object (ex, `]),`)
            oneline_state = state.clone(clone_shared=True)
            oneline_state.stream = TextIO()
            oneline_state.write_constrain = state.max_width - state.s.cur_line_length - 3
            try:
                self._write_nested_real(object, oneline_state, typeish, oneline=True)
                oneline_value = oneline_state.stream.getvalue()
                if "\n" in oneline_value:
                    oneline_value = None
            except oneline_state.WriteConstrained:
                oneline_value = None
            if oneline_value is not None:
                state.write(oneline_value)
                return
            state.write("\n" + state.get_indent_string())
            self._write_nested_real(object, state, typeish)
        finally:
            state.level -= 1
        state.write(state.get_indent_string())

    def _write_nested_real(self, object, state, typeish, oneline=False):
        indent_str = state.get_indent_string()
        first = True
        joiner = oneline and ", " or ",\n" + indent_str
        if typeish == "dict":
            items = object.items()
            if state.sort_dicts:
                items = _sorted(items)

            for k, v in items:
                if first:
                    first = False
                else:
                    state.write(joiner)
                self._format(k, state)
                state.write(": ")
                self._format(v, state)
        elif typeish == "odict":
            for k, v in object.items():
                if first:
                    first = False
                else:
                    state.write(joiner)
                state.write("(")
                self._format(k, state)
                state.write(", ")
                self._format(v, state)
                state.write(")")
        else:
            if typeish == "set":
                object = _sorted(object)
            for o in object:
                if first:
                    first = False
                else:
                    state.write(joiner)
                self._format(o, state)
        if oneline and typeish == "tuple" and len(object) == 1:
            state.write(", ")
        elif not oneline:
            state.write(",\n")

    def _format(self, object, state):
        write = state.write
        if state.max_depth and state.level >= state.max_depth:
            write("...")
            return
        state = state.clone()
        objid = id(object)
        if objid in state.context:
            write(self._recursion(object, state))
            return

        typ = type(object)
        r = typ.__repr__
        # Note: see comments on _mk_open_close_empty_dict for the rationale
        # behind looking up based first on type then on __repr__.
        try:
            opener_closer_empty = self._open_close_empty.get(typ) or self._open_close_empty.get(r)
        except TypeError:
            # This will happen if the type or the __repr__ is unhashable.
            # See: https://github.com/wolever/pprintpp/issues/18
            opener_closer_empty = None

        if opener_closer_empty is not None:
            orig_type, typeish, opener, closer, empty = opener_closer_empty
            if typ != orig_type:
                if opener is not None and "__PP_TYPE__" not in opener:
                    opener = "__PP_TYPE__(" + opener
                    closer = closer + ")"
                if empty is not None and "__PP_TYPE__" not in empty:
                    empty = "__PP_TYPE__(%s)" % (empty,)

            if r == defaultdict.__repr__:
                factory_repr = object.default_factory
                opener = "__PP_TYPE__(%r, {" % (factory_repr,)
                empty = opener + closer

            length = len(object)
            if length == 0:
                if "__PP_TYPE__" in empty:
                    empty = empty.replace("__PP_TYPE__", typ.__name__)
                write(empty)
                return

            if "__PP_TYPE__" in opener:
                opener = opener.replace("__PP_TYPE__", typ.__name__)
            write(opener)
            self._format_nested_objects(object, state, typeish=typeish)
            write(closer)
            return

        if r == bytes.__repr__:
            write(repr(object))
            return

        if r == str.__repr__:
            if str.isascii(object):  # Optimalization
                write(repr(object))
                return
            if "'" in object and '"' not in object:
                quote = '"'
                quotes = {'"': '\\"'}
            else:
                quote = "'"
                quotes = {"'": "\\'"}
            qget = quotes.get
            ascii_table_get = ascii_table.get
            unicat_get = unicodedata.category
            write(quote)
            for char in object:
                if ord(char) > 0x7F:
                    cat = unicat_get(char)
                    if unicode_printable_categories.get(cat):
                        try:
                            write(char)
                            continue
                        except UnicodeEncodeError:
                            pass
                write(qget(char) or ascii_table_get(char) or chr_to_ascii(char))
            write(quote)
            return

        orepr = repr(object)
        orepr = orepr.replace("\n", "\n" + state.get_indent_string())
        state.s.readable = state.s.readable and not orepr.startswith("<")
        write(orepr)
        return

    def _repr(self, object, context, level):
        repr, readable, recursive = self.format(object, context.copy(), self._depth, level)
        if not readable:
            self._readable = False
        if recursive:
            self._recursive = True
        return repr

    def format(self, object, context, maxlevels, level):
        """Format object for a specific context, returning a string
        and flags indicating whether the representation is 'readable'
        and whether the object represents a recursive construct.
        """
        state = self.get_default_state()
        result = self.pformat(object, state=state)
        return result, state.s.readable, state.s.recursive

    def _recursion(self, object, state):
        state.s.recursive = True
        return "<Recursion on %s with id=%s>" % (type(object).__name__, id(object))


if __name__ == "__main__":
    try:
        import numpy as np
    except ImportError:

        class np(object):  # type: ignore
            @staticmethod
            def array(o):
                return o

    somelist = [1, 2, 3]
    recursive = []  # type: ignore  # TODO: Add type hinting for mypy
    recursive.extend([recursive, recursive, recursive])
    pprint(
        {
            "a": {"a": "b"},
            "b": [somelist, somelist],
            "c": [
                (1,),
                (1, 2, 3),
            ],
            "ordereddict": OrderedDict(
                [
                    (1, 1),
                    (10, 10),
                    (2, 2),
                    (11, 11),
                ]
            ),
            "counter": [
                Counter(),
                Counter("asdfasdfasdf"),
            ],
            "dd": [
                defaultdict(int, {}),
                defaultdict(int, {"foo": 42}),
            ],
            "frozenset": frozenset("abc"),
            "np": [
                "hello",
                # np.array([[1,2],[3,4]]),
                "world",
            ],
            "u": ["a", "\u1234", "b"],
            "recursive": recursive,
            "z": {
                "very very very long key stuff 1234": {
                    "much value": "very nest! " * 10,
                    "unicode": "4U!'\"",
                },
                "aldksfj alskfj askfjas fkjasdlkf jasdlkf ajslfjas": ["asdf"] * 10,
            },
        }
    )
    pprint("\xe9e\u0301")
    uni_safe = "\xe9 \u6f02 \u0e4f \u2661"
    uni_unsafe = "\u200a \u0301 \n"
    unistr = uni_safe + " --- " + uni_unsafe
    sys.modules.pop("locale", None)
    pprint(unistr)
    stream = TextIO(encoding="ascii")
    pprint(unistr, stream=stream)
    print(stream.getvalue())


def load_ipython_extension(ipython):
    from .ipython import load_ipython_extension

    return load_ipython_extension(ipython)


def unload_ipython_extension(ipython):
    from .ipython import unload_ipython_extension

    return unload_ipython_extension(ipython)
